File: audiobook/src/main/java/de/ph1b/audiobook/uitools/PlayPauseDrawable.java
package de.ph1b.audiobook.uitools;
/*
 * Copyright (C) 2014 AChep@xda <artemchep@gmail.com>
 * Modified by Paul Woitaschek <woitaschek@posteo.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PixelFormat;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.util.Property;


/**
 * A class for creating simple transformation buttons. It is very simple to
 * use and perfectly fits simple Material icons' transformation.
 *
 * @author Artem Chepurnoy
 */
public class PlayPauseDrawable extends Drawable {


    /**
     * Pause icon
     */
    private static final float[][] VERTEX_PAUSE = {
            {10f, 6f, 6f, 10f, 10f, 14f, 14f, 18f, 18f},
            {5f, 5f, 19f, 19f, 5f, 5f, 19f, 19f, 5f}
    };

    /**
     * Play icon
     */
    private static final float[][] VERTEX_PLAY = {
            {19f, 19f, 8f, 8f, 19f, 19f, 8f, 8f, 19f},
            {12f, 12f, 5f, 9f, 12f, 12f, 5f, 19f, 12f}
    };
    private final Animator animator = ObjectAnimator.ofFloat(this, TRANSFORM, 0f, 1f);
    private final Path path;
    private final Paint paint;
    private final float[][][] vertex;
    private float progress;
    private int fromShape;
    private int toShape;
    private final static Property<PlayPauseDrawable, Float> TRANSFORM =
            new FloatProperty<PlayPauseDrawable>() {
                @Override
                public void setValue(PlayPauseDrawable object, float value) {
                    object.setTransformation(value);
                }

                @Override
                public Float get(PlayPauseDrawable object) {
                    return object.getTransformation();
                }
            };

    public PlayPauseDrawable() {
        this(VERTEX_PAUSE, VERTEX_PLAY);
    }

    private PlayPauseDrawable(@NonNull float[][]... vertex) {
        this.vertex = vertex;

        path = new Path();
        path.setFillType(Path.FillType.WINDING);

        paint = new Paint();
        paint.setAntiAlias(true);
        paint.setColor(Color.WHITE);
        paint.setStyle(Paint.Style.FILL);
    }

    public void transformToPause(boolean animated) {
        transformToShape(0, animated);
    }

    public void transformToPlay(boolean animated) {
        transformToShape(1, animated);
    }

    /**
     * public void setColor(int color) {
     * paint.setColor(color);
     * invalidateSelf();
     * }*
     */

    private void transformToShape(int i, boolean animated) {
        // Otherwise this will not be animated.
        if (toShape != i) {
            if (animated) {
                animator.setDuration(300);
            } else {
                animator.setDuration(0);
            }

            setTransformationTarget(i);
            animator.cancel();
            animator.start();
        }
    }

    private void setTransformationTarget(int i) {
        fromShape = toShape;
        toShape = i;
    }

    private float getTransformation() {
        return progress;
    }

    private void setTransformation(float progress) {
        this.progress = progress;
        Rect rect = getBounds();

        final float size = Math.min(rect.right - rect.left, rect.bottom - rect.top);
        final float left = rect.left + (rect.right - rect.left - size) / 2;
        final float top = rect.top + (rect.bottom - rect.top - size) / 2;

        path.reset();
        path.moveTo(
                left + calcTransformation(0, 0, progress, size),
                top + calcTransformation(1, 0, progress, size));
        for (int i = 1; i < vertex[0][0].length; i++) {
            path.lineTo(
                    left + calcTransformation(0, i, progress, size),
                    top + calcTransformation(1, i, progress, size));
        }

        path.close();
        invalidateSelf();
    }

    private float calcTransformation(int type, int i, float progress, float size) {
        float v0 = vertex[fromShape][type][i] * (1f - progress);
        float v1 = vertex[toShape][type][i] * progress;
        return (v0 + v1) * size / 24f;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void onBoundsChange(Rect bounds) {
        super.onBoundsChange(bounds);
        setTransformation(progress);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void draw(Canvas canvas) {
        canvas.drawPath(path, paint);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setAlpha(int alpha) {
        paint.setAlpha(alpha);
        invalidateSelf();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setColorFilter(ColorFilter cf) {
        paint.setColorFilter(cf);
        invalidateSelf();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public int getOpacity() {
        return PixelFormat.TRANSLUCENT;
    }

}

File: app/src/main/java/de/ph1b/audiobook/uitools/PlayPauseDrawable.kt
package de.ph1b.audiobook.uitools

import android.animation.Animator
import android.animation.ObjectAnimator
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ColorFilter
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PixelFormat
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.util.Property
import android.view.animation.DecelerateInterpolator

/**
 * This code was modified by me, Paul Woitaschek. All these changes are licensed under GPLv3. The
 * original source can be found here: {@see https://github.com/alexjlockwood/material-pause-play-
 * * animation/blob/master/app/src/main/java/com/alexjlockwood/example/playpauseanimation/
 * * PlayPauseView.java}
 *
 * The original licensing is as follows:
 *
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 Alex Lockwood
 *
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and
 * associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 *
 * The above copyright notice and this permission notice shall be included in all copies or
 * substantial portions of the Software.
 *
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
class PlayPauseDrawable : Drawable() {

  private var pauseBarHeight = 0F
  private var pauseBarWidth = 0F
  private var pauseBarDistance = 0F
  private val leftPauseBar = Path()
  private val rightPauseBar = Path()
  private val interpolator = DecelerateInterpolator()
  private val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.FILL
    color = Color.WHITE
  }
  private var progress: Float = 0F
    set(progress) {
      field = progress
      invalidateSelf()
    }
  private val progressProperty = object : Property<PlayPauseDrawable, Float>(Float::class.java, "progress") {
    override fun get(d: PlayPauseDrawable) = d.progress
    override fun set(d: PlayPauseDrawable, value: Float) {
      d.progress = value
    }
  }
  private var isPlay: Boolean = false
  private var animator: Animator? = null

  /** Linear interpolate between a and b with parameter t.  */
  private fun interpolate(a: Float, b: Float, t: Float) = a + (b - a) * t

  override fun onBoundsChange(bounds: Rect) {
    super.onBoundsChange(bounds)

    pauseBarHeight = 7.0f / 12.0f * (bounds.height().toFloat())
    pauseBarWidth = pauseBarHeight / 3.0f
    pauseBarDistance = pauseBarHeight / 3.6f
  }

  override fun draw(canvas: Canvas) {
    leftPauseBar.rewind()
    rightPauseBar.rewind()

    // move to center of canvas
    canvas.translate(bounds.left.toFloat(), bounds.top.toFloat())

    // The current distance between the two pause bars.
    val barDist = interpolate(pauseBarDistance, 0.0f, progress) - 1 // -1 because else there is a small gap on some devices
    // The current width of each pause bar.
    val barWidth = interpolate(pauseBarWidth, pauseBarHeight / 1.75f, progress)
    // The current position of the left pause bar's top left coordinate.
    val firstBarTopLeft = interpolate(0.0f, barWidth, progress)
    // The current position of the right pause bar's top right coordinate.
    val secondBarTopRight = interpolate(2.0f * barWidth + barDist, barWidth + barDist, progress)

    // Draw the left pause bar. The left pause bar transforms into the
    // top half of the play button triangle by animating the position of the
    // rectangle's top left coordinate and expanding its bottom width.
    leftPauseBar.moveTo(0.0f, 0.0f)
    leftPauseBar.lineTo(firstBarTopLeft, -pauseBarHeight)
    leftPauseBar.lineTo(barWidth, -pauseBarHeight)
    leftPauseBar.lineTo(barWidth, 0.0f)
    leftPauseBar.close()

    // Draw the right pause bar. The right pause bar transforms into the
    // bottom half of the play button triangle by animating the position of the
    // rectangle's top right coordinate and expanding its bottom width.
    rightPauseBar.moveTo(barWidth + barDist, 0.0f)
    rightPauseBar.lineTo(barWidth + barDist, -pauseBarHeight)
    rightPauseBar.lineTo(secondBarTopRight, -pauseBarHeight)
    rightPauseBar.lineTo(2.0f * barWidth + barDist, 0.0f)
    rightPauseBar.close()

    canvas.save()

    // Translate the play button a tiny bit to the right so it looks more centered.
    canvas.translate(interpolate(0.0f, pauseBarHeight / 8.0f, progress), 0.0f)

    // (1) Pause --> Play: rotate 0 to 90 degrees clockwise.
    // (2) Play --> Pause: rotate 90 to 180 degrees clockwise.
    val rotationProgress = if (isPlay) 1.0f - progress else progress
    val startingRotation = if (isPlay) 90.0f else 0.0f
    canvas.rotate(interpolate(startingRotation, startingRotation + 90.0f, rotationProgress), bounds.width() / 2.0f, bounds.height() / 2.0f)

    // Position the pause/play button in the center of the drawable's bounds.
    canvas.translate(bounds.width() / 2.0f - ((2.0f * barWidth + barDist) / 2.0f), bounds.height() / 2.0f + (pauseBarHeight / 2.0f))

    // Draw the two bars that form the animated pause/play button.
    canvas.drawPath(leftPauseBar, paint)
    canvas.drawPath(rightPauseBar, paint)

    canvas.restore()
  }

  fun transformToPause(animated: Boolean) {
    if (isPlay) {
      jumpToCurrentState()
      if (animated) {
        toggle()
      } else {
        isPlay = false
        progress = 0.0f
        invalidateSelf()
      }
    }
  }

  fun transformToPlay(animated: Boolean) {
    if (!isPlay) {
      jumpToCurrentState()
      if (animated) {
        toggle()
      } else {
        isPlay = true
        progress = 1.0f
        invalidateSelf()
      }
    }
  }

  override fun jumpToCurrentState() {
    animator?.cancel()
    progress = if (isPlay) 1.0f else 0.0f
  }

  private fun toggle() {
    animator?.cancel()
    val from = if (isPlay) 1.0f else 0.0f
    val to = if (isPlay) 0.0f else 1.0f
    isPlay = !isPlay
    animator = ObjectAnimator.ofFloat(this, progressProperty, from, to)
        .also {
          it.interpolator = interpolator
          it.duration = 275
          it.start()
        }
  }

  override fun setAlpha(alpha: Int) {
    paint.alpha = alpha
    invalidateSelf()
  }

  override fun setColorFilter(cf: ColorFilter?) {
    paint.colorFilter = cf
    invalidateSelf()
  }

  override fun getOpacity() = PixelFormat.TRANSLUCENT
}
